wasm/jit/jit_amd64.go

```diff
@@ -1122,6 +1122,228 @@ func (c *amd64Compiler) compilePopcnt(o *wazeroir.OperationPopcnt) error {
+		return err
+	}
+	// Now we have the quotient of the division result in the AX register,
+	// so we record it.
+	result := c.locationStack.pushValueOnRegister(x86.REG_AX)
+	result.setRegisterType(generalPurposeRegisterTypeInt)
+
+	// Make sure that the DX register (holding remainder of the division) is marked unused.
+	c.locationStack.markRegisterUnused(x86.REG_DX)
+	return nil
+}
+
+// compileRem emits the instructions to perform division on the top
+// two values of integer type on the stack and put the remainder of the result
+// onto the stack. For example, stack [..., 8, 3] results in [..., 2] where
+// the quotient is discarded. See compileDivForInts for how to acquire quotient, not remainder.
+func (c *amd64Compiler) compileRem(o *wazeroir.OperationRem) (err error) {
+	switch o.Type {
+	case wazeroir.SignedInt32:
+		err = c.performDivisionOnInts(true, true)
+	case wazeroir.SignedInt64:
+		err = c.performDivisionOnInts(false, true)
+	case wazeroir.SignedUint32:
+		err = c.performDivisionOnInts(true, false)
+	case wazeroir.SignedUint64:
+		err = c.performDivisionOnInts(false, false)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Now we have the remainder of the division result in the DX register,
+	// so we record it.
+	result := c.locationStack.pushValueOnRegister(x86.REG_DX)
+	result.setRegisterType(generalPurposeRegisterTypeInt)
+
+	// Make sure that the AX register (holding quotient of the division) is marked unused.
+	c.locationStack.markRegisterUnused(x86.REG_AX)
+	return
+}
+
+// performDivisionOnInts emits the instructions to do divisions on top two integers on the stack
+// via DIV (unsigned div) and IDIV (signed div) instructions.
+// See the following explanation of these instructions' semantics from https://www.lri.fr/~filliatr/ens/compil/x86-64.pdf
+//
+// >> Division requires special arrangements: idiv (signed) and div (unsigned) operate on a 2n-byte dividend and
+// >> an n-byte divisor to produce an n-byte quotient and n-byte remainder. The dividend always lives in a fixed pair of
+// >> registers (%edx and %eax for the 32-bit case; %rdx and %rax for the 64-bit case); the divisor is specified as the
+// >> source operand in the instruction. The quotient goes in %eax (resp. %rax); the remainder in %edx (resp. %rdx). For
+// >> signed division, the cltd (resp. ctqo) instruction is used to prepare %edx (resp. %rdx) with the sign extension of
+// >> %eax (resp. %rax). For example, if a,b, c are memory locations holding quad words, then we could set c = a/b
+// >> using the sequence: movq a(%rip), %rax; ctqo; idivq b(%rip); movq %rax, c(%rip).
+//
+// tl;dr is that the division result is placed in AX and DX registers after instructions emitted by this function
+// where AX is holding the quotiend while DX holding remainder of the division result.
+func (c *amd64Compiler) performDivisionOnInts(is32Bit bool, signed bool) error {
+
+	const (
+		quotientRegister  = x86.REG_AX
```

Comment:
I like this, but also notice we often have to refer explicitly to the AX or DX... so not sure it hurts, but it is what it is.